sympy/tensor/array/array_comprehension.py

```diff
@@ -30,184 +31,250 @@ class ArrayComprehension(Basic):
-        for var, inf, sup in self._limits:
+        expr_free_sym = self.function.free_symbols
+        for var, inf, sup in self.limits:
             expr_free_sym.discard(var)
-            if len(inf.free_symbols) > 0:
+            if inf.free_symbols:
                 expr_free_sym = expr_free_sym.union(inf.free_symbols)
-            if len(sup.free_symbols) > 0:
+            if sup.free_symbols:
                 expr_free_sym = expr_free_sym.union(sup.free_symbols)
         return expr_free_sym
 
     @property
     def variables(self):
-        """
-        Return a list of the variables in the limits
+        """Used for obtaining the tuples of the variables in the limits
+
+        Returns
+        =======
+
+        tuple
+            The variables in the limits
 
         Examples
         ========
 
         >>> from sympy.tensor.array import ArrayComprehension
-        >>> from sympy.abc import i, j, k
+        >>> from sympy import symbols
+        >>> i, j, k = symbols('i j k')
         >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
         >>> a.variables
         [i, j]
         """
-        return [l[0] for l in self._limits]
+        return [l[0] for l in self.limits]
 
     @property
     def bound_symbols(self):
+        """Used to obtain the dummay variables
+
+        Returns
+        =======
+
+        list
+            The list of bound symbols
+
+        Note
+        ====
+
+        Note that all variables are dummy variables since a limit without
+        lower bound or upper bound is not accepted.
         """
-        Return only variables that are dummy variables. Note that all variables are
-        dummy variables since a limit without lower bound or upper bound is not accpted.
-        """
-        return [l[0] for l in self._limits if len(l) != 1]
+        return [l[0] for l in self.limits if len(l) != 1]
 
     @property
     def shape(self):
         """
-        Return the shape of the expanded array, which can have symbols. Note that both
-        the lower and the upper bounds are included while calculating the shape.
+
+        Used for obtaining the shape of the expanded array, which may have symbols
+
+        Returns
+        =======
+
+        tuple
+            The shape of the expanded array.
+
+        Note
+        ====
+
+        Both the lower and the upper bounds are included while
+        calculating the shape.
 
         Examples
         ========
 
         >>> from sympy.tensor.array import ArrayComprehension
-        >>> from sympy.abc import i, j, k
+        >>> from sympy import symbols
+        >>> i, j, k = symbols('i j k')
         >>> a = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, 3))
         >>> a.shape
         (4, 3)
         >>> b = ArrayComprehension(10*i + j, (i, 1, 4), (j, 1, k+3))
         >>> b.shape
         (4, k + 3)
         """
-        return self._shape
+        return self._calculate_shape_from_limits(self.limits)
```

Comment:
This will slow down the code quite a lot. I would prefer to maintain `self._shape`. The purpose of the underscore is to hide the field from users, so that it is unlikely to be modified.

There is still a risk that a user manually overwrites `_shape`, but I don't care. Using the code in a non-standard way entails possible anomalous behavior.